In [1]:
import pandas as pd 
df = pd.read_csv(r"C:\Users\Rasulbek907\Desktop\Project_One\Data\Raw_Data\titanic.csv")

In [2]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [5]:
def missing_val(df):
    for col in df.columns:
        if df[col].isnull().sum():
            if df[col].dtype == "object":
                df[col] = df[col].fillna(df[col].mode()[0])
            else:
                df[col] = df[col].fillna(df[col].mean())
    return df 

In [ ]:
df = missing_val(df)

In [7]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [ ]:
df.to_csv(r"C:\Users\Rasulbek907\Desktop\Project_One\Data\Preprocessed_Data\Preprocessed_Missing_value.csv")

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        891 non-null    object 
 11  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
import pandas as pd 
from sklearn.preprocessing import LabelEncoder


def encoder_cat(df, one_hot_threshold=6):
    encoder = LabelEncoder()

    for col in df.columns:
        if df[col].dtype=="object":
            if df[col].nunique() <= one_hot_threshold:
                dummies = pd.get_dummies(df[col], prefix="col", dtype=int)
                df = pd.concat([df.drop(columns=[col]), dummies], axis=1)

            else:
                df[col] = encoder.fit_transform(df[col])
    return df 

In [10]:
df = encoder_cat(df)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    int64  
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       891 non-null    int64  
 8   Fare         891 non-null    float64
 9   Cabin        891 non-null    int64  
 10  col_female   891 non-null    int64  
 11  col_male     891 non-null    int64  
 12  col_C        891 non-null    int64  
 13  col_Q        891 non-null    int64  
 14  col_S        891 non-null    int64  
dtypes: float64(2), int64(13)
memory usage: 104.5 KB


In [12]:
df.head()

,PassengerId,Survived,Pclass,Name,Age,SibSp,Parch,Ticket,Fare,Cabin,col_female,col_male,col_C,col_Q,col_S
0,1,0,3,108,22.0,1,0,523,7.2500,47,0,1,0,0,1
1,2,1,1,190,38.0,1,0,596,71.2833,81,1,0,1,0,0
2,3,1,3,353,26.0,0,0,669,7.9250,47,1,0,0,0,1
3,4,1,1,272,35.0,1,0,49,53.1000,55,1,0,0,0,1
4,5,0,3,15,35.0,0,0,472,8.0500,47,0,1,0,0,1


In [ ]:
df.to_csv(r"C:\Users\Rasulbek907\Desktop\Project_One\Data\Preprocessed_Data\Preprocessed_Encoding.csv")

# Standart Scaling 

In [13]:
from sklearn.preprocessing import StandardScaler

num_col = [
    'Age', "SibSp", "Parch", "Fare", 'Pclass', "col_female", "col_male", 'col_C', 'col_Q', "col_S", "Cabin", "Ticket", "Name"
]

scaler = StandardScaler()

df[num_col] = scaler.fit_transform(df[num_col])

In [14]:
df.head()

,PassengerId,Survived,Pclass,Name,Age,SibSp,Parch,Ticket,Fare,Cabin,col_female,col_male,col_C,col_Q,col_S
0,1,0,0.827377,-1.310217,-0.592481,0.432793,-0.473674,0.918966,-0.502445,-0.281881,-0.737695,0.737695,-0.482043,-0.307562,0.615838
1,2,1,-1.566107,-0.991410,0.638789,0.432793,-0.473674,1.282625,0.786845,1.161545,1.355574,-1.355574,2.074505,-0.307562,-1.623803
2,3,1,0.827377,-0.357685,-0.284663,-0.474545,-0.473674,1.646283,-0.488854,-0.281881,1.355574,-1.355574,-0.482043,-0.307562,0.615838
3,4,1,-1.566107,-0.672604,0.407926,0.432793,-0.473674,-1.442322,0.420730,0.057748,1.355574,-1.355574,-0.482043,-0.307562,0.615838
4,5,0,0.827377,-1.671790,0.407926,-0.474545,-0.473674,0.664904,-0.486337,-0.281881,-0.737695,0.737695,-0.482043,-0.307562,0.615838


In [16]:
df.to_csv(r"C:\Users\Rasulbek907\Desktop\Project_One\Data\Preprocessed_Data\Preprocessed_Data.csv")

In [1]:
import pandas as pd 
df = pd.read_csv(r"C:\Users\Rasulbek907\Desktop\Project_One\Data\Preprocessed_Data\Preprocessed_Data.csv")

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   891 non-null    int64  
 1   PassengerId  891 non-null    int64  
 2   Survived     891 non-null    int64  
 3   Pclass       891 non-null    float64
 4   Name         891 non-null    float64
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    float64
 7   Parch        891 non-null    float64
 8   Ticket       891 non-null    float64
 9   Fare         891 non-null    float64
 10  Cabin        891 non-null    float64
 11  col_female   891 non-null    float64
 12  col_male     891 non-null    float64
 13  col_C        891 non-null    float64
 14  col_Q        891 non-null    float64
 15  col_S        891 non-null    float64
dtypes: float64(13), int64(3)
memory usage: 111.5 KB


In [3]:
df.head()

,Unnamed: 0,PassengerId,Survived,Pclass,Name,Age,SibSp,Parch,Ticket,Fare,Cabin,col_female,col_male,col_C,col_Q,col_S
0,0,1,0,0.827377,-1.310217,-0.592481,0.432793,-0.473674,0.918966,-0.502445,-0.281881,-0.737695,0.737695,-0.482043,-0.307562,0.615838
1,1,2,1,-1.566107,-0.991410,0.638789,0.432793,-0.473674,1.282625,0.786845,1.161545,1.355574,-1.355574,2.074505,-0.307562,-1.623803
2,2,3,1,0.827377,-0.357685,-0.284663,-0.474545,-0.473674,1.646283,-0.488854,-0.281881,1.355574,-1.355574,-0.482043,-0.307562,0.615838
3,3,4,1,-1.566107,-0.672604,0.407926,0.432793,-0.473674,-1.442322,0.420730,0.057748,1.355574,-1.355574,-0.482043,-0.307562,0.615838
4,4,5,0,0.827377,-1.671790,0.407926,-0.474545,-0.473674,0.664904,-0.486337,-0.281881,-0.737695,0.737695,-0.482043,-0.307562,0.615838


In [4]:
x = df.drop(columns=["Survived"])
y = df["Survived"] 

In [6]:
from sklearn.model_selection import train_test_split
x_train, x_temp, y_train, y_temp = train_test_split (x, y, test_size=0.3, random_state=42, stratify=y)

In [7]:
x_val, x_test, y_val, y_test = train_test_split(x_temp, y_temp, test_size=0.5, random_state=42, stratify=y_temp)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


model = LogisticRegression(max_iter=1000, random_state=42)
model.fit(x_train, y_train)
y_val_pred = model.predict(x_val)
val_score = accuracy_score(y_val, y_val_pred)
print("Validation Accuracy Score :  ", val_score)

Validation Accuracy Score :   0.8283582089552238


In [17]:
y_true = y_test
y_pred = model.predict(x_test)
test_accuracy = accuracy_score(y_true, y_pred)

print("Test Accuracy Score: ", test_accuracy)

Test Accuracy Score:  0.746268656716418


# Classification modellar uchun METRICS

# Confusion Metrix - Bu eng asosiy metrik hisoblanadi

In [19]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_true, y_pred)
print(cm)

[[67 16]
 [18 33]]


# CM 4 ta qismga bo`linib natijalarni beradi
- 1-TP - True Positiv 
- TN - True Negative  
- 3-FP - False Positiv  
- 4-FN - False  Negative 

- TP - Spam bo`lgan emailni spam deb togri topdi 
- TN - Spam bo`lmagan emailni spam emas deb tog`ri belgiladi 

- FP - spam emas bo`lgan emailni spam deb xato belgiladi 
- FN - spam bo`lgan emailni spam emasdeb xato belgiladi

# Accuracy Metrics 

# Accuracy - bu modelning umumiy tog`ri bashoratlari ulushi

# Misol UChun : Test Accuracy Score:  0.746268656716418

In [20]:
from sklearn.metrics import accuracy_score
accuracy_score(y_true, y_pred)

0.746268656716418

# Precision - Metrics

# Precision - bu model positive deganlarning nechtasi haqiqatdan positive

# Real Misol: Spam qiluvchi model : model spam dedi - lekin bu haqiqiy spam email bo`lmasligi mumkin 

In [22]:
from sklearn.metrics import precision_score
precision_score(y_true, y_pred)

0.673469387755102

# Recall Metrics

# Recall - bu haqqiqiy positivlarni qanchalik yaxshi topa olishi 
# real misol: firibgarlik bor edi shu firibgarlikni qanchalik o`tkazib yubordi yoki ushlay oldi: 
# jinoyatchini topuvchu SI modeli haqiqatdan o`tkazib yubormay topa oldimi`

In [24]:
from sklearn.metrics import recall_score
recall_score(y_true, y_pred)

0.6470588235294118

# F1 Score - Metrics

# f1 - bu precision va recallning garmonik o`rtachasi // ikkalasining muvozanatini beruvchi metrik

In [26]:
from sklearn.metrics import f1_score
f1_score(y_true, y_pred)

0.66

# Roc Auc - 
- Roc Curve  x o`qi FPR(False positive rate) // y o`qi (TPR- true positive rate)

- Auc - roc egri chizigi ostidagi maydon - 0.5 random // 0.7-0.8 yaxshi // 0.9 + juda yaxshi 

In [27]:
from sklearn.metrics import roc_auc_score
roc_auc_score(y_true, y_pred)

0.7271438695960312

# Qachon qaysi Metrikni tanlash kerak 

# Vaziyat: Balanslangan Dataset   Eng Yaxshi Metrik: Accuracy
# Vaziyat: FP zarur bo`lsa   Eng Yaxshi Metrik: Precision 
# Vaziyat:   FN zarur bo`lsa  Eng Yaxshi Metrik: Recall
# Vaziyat: Imbalanced dataset   Eng Yaxshi Metrik: f1-score
# Vaziyat: probabilityni baholash   Eng Yaxshi Metrik: ROC-AUC

# REGRESSION MODELLAR UCHUN METRIKLAR

# metriklar - MAE // MSE // RMSE // R2 

# MAE - Mean Absolute Error -- model o`rtacha qancha xato qilyabdi

In [28]:
from sklearn.metrics import mean_absolute_error

mae = mean_absolute_error(y_true, y_pred)

print("MAE: ", mae)

MAE:  0.2537313432835821


# MSE - Mean Squared Error --  katta xatolarni qattiq jazolaydi 

In [29]:
from sklearn.metrics import mean_squared_error

mse = mean_squared_error(y_true, y_pred)

print("MSE: ", mse)

MSE:  0.2537313432835821


# RMSE - Root Mean Squared Error - ildiz ostidan natijani chiqarish

In [31]:
import numpy as np
rmse = np.sqrt(mse)
print("RMSE: ", rmse)

RMSE:  0.5037175233040658


# R2 - R Squared - coeffitsient of determination

# Model targetni nechchi % tushuntira oldi // 1-mukammal // 0-foydasiz // - manfiy yamon

| Metrik | Savolga javob                          |
| ------ | -------------------------------------- |
| MAE    | O‘rtacha xato qancha?                  |
| MSE    | Katta xatolar qanchalik yomon?         |
| RMSE   | Real xato miqdori qancha?              |
| R²     | Model qanchalik yaxshi tushuntiryapti? |


# Qachon qaysi metrik ishlatiladi 
- Regression umumiy natijasi muhim bo`lsa  -- MAE 
- Katta xatolar xavfi b`olsa  - RMSE 
- Modelni tushuntirish darajasi kerak bo`ladigan  bo`lsa  --  R2 

# Classification Metrics - Accuracy/ precision / Recall / f1 score/ roc auc 
# Regression Metrics - MAE/ MSE / RMSE / R2 